lib/src/extension_tree.dart

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A structure to keep track of extensions and their applicability efficiently.
+library dartdoc.extension_tree;
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:dartdoc/src/element_type.dart';
+import 'package:dartdoc/src/model.dart';
+import 'package:dartdoc/src/special_elements.dart';
+
+/// This class defines a node in an [Extension] tree.  A hybrid of a heap and
+/// a type tree, an extension tree's nodes are arranged to the following
+/// invariants on public methods:
+///
+/// - Each node contains the set of extensions declared as being directly on
+///   [extendedType], or an equivalent type.
+/// - All parents, recursively, of a node and the node itself contain extensions
+///   that could apply to any instantiated type of [extendedType].
+///
+/// An extension "could apply" to a type if there exists some valid
+/// instantiation of that type where the extension would be applicable if
+/// made accessible in the namespace per:
+/// https://github.com/dart-lang/language/blob/master/accepted/2.6/static-extension-members/feature-specification.md#implicit-extension-member-invocation
+class ExtensionNode {
```

Comment:
I'm not convinced that a tree structure is adequate, given that it appears to follow the type hierarchy but the type hierarchy is a graph structure. If the declarations look like
```
class A {}
class B extends A {}
```
then, as I understand it, there would be a node for `A` and a node for `B` that is a child of the node for `A`. But if the declarations look like
```
class A {}
class B {}
class C implements A, B {}
```
then there would be nodes for `A`, `B`, and `C`, and `C` would need to be a child of both the node for `A` and the node for `B`, which a tree doesn't allow.

In analyzer we just compute a list of the extensions visible within a single library, then recompute the set of applicable extensions based on a given type without trying to build a graph of extensions. I think that's a cleaner approach, but ymmv.